File: 16_Intervals/01_Insert_Interval/0057-insert-interval.py
"""
Problem: LeetCode 57 - Insert Interval

Key Idea:
We can solve this problem by iterating through the intervals and keeping track of the merged intervals. We initialize an empty list `result` to store the merged intervals. We iterate through the intervals, and for each interval, we compare its end with the start of the new interval and the new interval's end with the start of the current interval. If they overlap, we update the start and end of the new interval accordingly. If they don't overlap, we add the current interval to the result and reset the new interval. After iterating through all intervals, we add the new interval (if not added already) and return the result.

Time Complexity:
- The time complexity of this approach is O(n), where n is the number of intervals.

Space Complexity:
- The space complexity is O(n), as we store the merged intervals in the result list.
"""


class Solution:
    def insert(
        self, intervals: List[List[int]], newInterval: List[int]
    ) -> List[List[int]]:
        result = []
        new_start, new_end = newInterval

        for interval in intervals:
            if interval[1] < new_start:
                result.append(interval)
            elif interval[0] > new_end:
                result.append([new_start, new_end])
                new_start, new_end = interval
            else:
                new_start = min(new_start, interval[0])
                new_end = max(new_end, interval[1])

        result.append([new_start, new_end])
        return result

File: 16_Intervals/01_Insert_Interval/0057-insert-interval.cpp
/*
Problem: LeetCode 57 - Insert Interval

Description:
Given a set of non-overlapping intervals sorted by their start times,
insert a new interval into the intervals (merge if necessary).

You may assume that the intervals were initially sorted according to their start times.

Intuition:
The key idea is to find the correct position to insert the new interval and merge any overlapping intervals if necessary.

Approach:
1. Create an empty result vector to store the merged intervals.
2. Iterate through the given intervals:
   a. If the current interval's end is less than the new interval's start, add the current interval to the result vector.
   b. If the current interval's start is greater than the new interval's end, it means we have found the correct position to insert the new interval.
      Add the new interval to the result vector and merge any remaining intervals if needed.
   c. If there is an overlap between the current interval and the new interval, update the new interval's start and end to cover the overlapping region.
3. Add any remaining intervals to the result vector if needed.
4. Return the result vector.

Time Complexity:
The time complexity is O(n), where n is the number of intervals.

Space Complexity:
The space complexity is O(1) for the result vector, and no additional data structures are used, so the overall space complexity is also O(1).
*/

class Solution {
  public:
    vector<vector<int>> insert(vector<vector<int>> &intervals, vector<int> &newInterval) {
        vector<vector<int>> result;
        int i = 0;
        int n = intervals.size();

        // Add intervals that end before the new interval starts
        while (i < n && intervals[i][1] < newInterval[0]) {
            result.push_back(intervals[i]);
            i++;
        }

        // Merge overlapping intervals
        while (i < n && intervals[i][0] <= newInterval[1]) {
            newInterval[0] = min(newInterval[0], intervals[i][0]);
            newInterval[1] = max(newInterval[1], intervals[i][1]);
            i++;
        }

        // Add the merged interval
        result.push_back(newInterval);

        // Add remaining intervals
        while (i < n) {
            result.push_back(intervals[i]);
            i++;
        }

        return result;
    }
};